src/driver/amdxdna/aie2_event_trace.c

```diff
@@ -0,0 +1,368 @@
+	if (ret) {
+		XDNA_ERR(xdna, "Failed to register irq %d ret %d", msi_idx, ret);
+		goto destroy_wq;
+	}
+
+	req_buf->kern_log_buf = kcalloc(TRACE_EVENT_BUF_SIZE, sizeof(u8), GFP_KERNEL);
+	if (!req_buf->kern_log_buf) {
+		ret = -ENOMEM;
+		goto free_irq;
+	}
+	return 0;
+
+free_irq:
+	free_irq(req_buf->log_ch_irq, ndev);
+destroy_wq:
+	destroy_workqueue(req_buf->wq);
+free_dma_trace_buf:
+	dma_free_noncoherent(xdna->ddev.dev, req_buf->dram_buffer_size, req_buf->buf,
+			     (dma_addr_t)req_buf->dram_buffer_address, DMA_BIDIRECTIONAL);
+	req_buf->buf = NULL;
+	req_buf->dram_buffer_address = 0;
+	req_buf->dram_buffer_size = 0;
+	return ret;
+}
+
+static void aie2_deregister_log_buf_irq_hdl(struct amdxdna_dev_hdl *ndev)
+{
+	struct event_trace_req_buf *req_buf = ndev->event_trace_req;
+
+	cancel_work_sync(&req_buf->work);
+	/* print already accumulated FW logs */
+	aie2_print_trace_event_log(ndev);
+	destroy_workqueue(req_buf->wq);
+
+	free_irq(req_buf->log_ch_irq, ndev);
+	kfree(req_buf->kern_log_buf);
+}
+
+static int aie2_event_trace_alloc(struct amdxdna_dev_hdl *ndev)
+{
+	struct event_trace_req_buf *req_buf = ndev->event_trace_req;
+	struct amdxdna_dev *xdna = ndev->xdna;
+
+	req_buf->buf = dma_alloc_noncoherent(xdna->ddev.dev, TRACE_EVENT_BUF_SIZE,
+					     (dma_addr_t *)&req_buf->dram_buffer_address,
+					     DMA_BIDIRECTIONAL, GFP_KERNEL);
+
+	if (!req_buf->buf)
+		return -ENOMEM;
+
+	req_buf->dram_buffer_size = TRACE_EVENT_BUF_SIZE;
+	XDNA_DBG(ndev->xdna, "Start event trace buf addr: 0x%llx size 0x%x",
+		 req_buf->dram_buffer_address, req_buf->dram_buffer_size);
+
+	return 0;
+}
+
+static void aie2_event_trace_free(struct amdxdna_dev_hdl *ndev)
+{
+	struct event_trace_req_buf *req_buf = ndev->event_trace_req;
+	struct amdxdna_dev *xdna = ndev->xdna;
+
+	dma_free_noncoherent(xdna->ddev.dev, req_buf->dram_buffer_size, req_buf->buf,
+			     (dma_addr_t)req_buf->dram_buffer_address, DMA_BIDIRECTIONAL);
+
+	req_buf->buf = NULL;
+	req_buf->dram_buffer_address = 0;
+	req_buf->dram_buffer_size = 0;
+}
+
+static int aie2_start_event_trace_send(struct amdxdna_dev_hdl *ndev)
+{
+	struct event_trace_req_buf *req_buf = ndev->event_trace_req;
+	struct amdxdna_dev *xdna = ndev->xdna;
+	int ret;
+
+	ret = aie2_event_trace_alloc(ndev);
+	if (ret) {
+		XDNA_ERR(xdna, "Failed to allocate log buffer, ret %d", ret);
+		return ret;
+	}
+
+	drm_WARN_ON(&xdna->ddev, !mutex_is_locked(&xdna->dev_lock));
+	ret = aie2_start_event_trace(ndev, req_buf->dram_buffer_address,
+				     req_buf->dram_buffer_size);
+	if (ret) {
+		XDNA_ERR(xdna, "Failed to start event trace, ret %d", ret);
+		aie2_event_trace_free(ndev);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int aie2_stop_event_trace_send(struct amdxdna_dev_hdl *ndev)
+{
+	struct amdxdna_dev *xdna = ndev->xdna;
+	int ret;
+
+	if (!ndev->event_trace_req) {
+		XDNA_DBG(xdna, "Event tracing is not started");
+		return 0;
+	}
+
```

Comment:
The condition will always be false. It was checked in aie2_assign_event_trace_state() and return early if this condition is true. But just a minor issue. Please delete this in your next PR.